// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";

package tensorflow.magenta;

import "magenta/protobuf/music.proto";

message GeneratorDetails {
  // A unique ID for the generator on this server.
  string id = 1;
  // A short, human-readable description of the generator.
  string description = 2;
}

// Options for generating a sequence.
message GeneratorOptions {
  message GenerateSection {
    // Start time for the model to fill, inclusive.
    double start_time = 1;
    // End time for the model to fill, inclusive.
    double end_time = 2;
    // Minimum 0-based MIDI pitch.
    int32 min_output_pitch = 3;
    // Maximum 0-based MIDI pitch.
    int32 max_output_pitch = 4;
  }

  repeated GenerateSection generate_sections = 1;
}

message ListGeneratorsRequest {
}

message ListGeneratorsResponse {
  repeated GeneratorDetails generators = 1;
}

message GenerateSequenceRequest {
  // An input sequence to base the generation on.
  NoteSequence input_sequence = 1;

  GeneratorOptions generator_options = 2;
}

message GenerateSequenceResponse {
  // The generated sequence.
  // This will always include the entire NoteSequence, including any notes that
  // occurred before and after the generated portion of the sequence.
  NoteSequence generated_sequence = 1;
}
